### ISSUE TITLE
Foreign Keyboard Management

### ISSUE THREAD

**Issue by T4ouf** [TheCherno/Hazel#261]
#### Description
Currently Hazel only "works" for QWERTY keyboards. I mean that input polling is incorrect for other type of keyboards (like AZERTY or QWERTZ keyboards). (I am not sure if this issue should be in 'bug' or in 'feature' section :/ )

#### To Reproduce
0. Plug a foreign keyboard and make sure it works properly when you type (meaning that it prints the key that is on the keyboard. Make sure that your OS is not overwritting the default keyboard keys). I am currently using a french AZERTY keyboard.
1. Go to SandboxApp.cpp
2. Paste the following code in the OnEvent(....) function : 
```cpp
if (event.GetEventType() == Hazel::EventType::KeyPressed) {
		Hazel::KeyPressedEvent& e = (Hazel::KeyPressedEvent&)event;
		if(e.GetKeyCode() == HZ_KEY_A)
			HZ_TRACE("{0}", char(e.GetKeyCode()));
		}
}
```
3. Run SandboxApp.cpp.
4. If you have an AZERTY keyboard (which is my case) : when you click on "A" (where "Q" is located on a QWERTY keyboard), it will not work. You need to press Q (where "A" is located on a QWERTY keyboard) to enter the "if" statement.

#### Expected behavior
In a "normal" case, I would expect the engine (or the underneath library) to handle correctly all type of keyboards. Pressing "A" should not mean pressing any other key than "A" (regardless of the keyboard type)

### Why is it a problem ?
It can lead to bugs in the usage of Hazel by foreign developpers (+ can mislead the users).

#### Operating system: 
- OS: win-64
- Version: Windows 10 professional

**Comment by BaderEddineOuaich** [comment]
This used to be an issue for non us keyboards since GLFW exists...
although in glfw 3.2, they provide glfwGetKeyName function. which can be used to solve this issue by mapping keys to be used by all keyboard types.
i agree with @T4ouf as this can be serious for non us keyboard users. and should be one of the priorities to implement.

**Comment by T4ouf** [comment]
I'll check glfwGetKeyName and see how to implement it correctly ! If it works, I'll make a pull request. (Yet I don't know how much work I'll be able to involve in this :/ Also, maybe it is a big change and it will need to be supervised by @TheCherno in a video or something ?)

**Comment by T4ouf** [comment]
I think I have found a solution to the problem ! I've made a pull request in order to fix this issue ! Thank you for your help @BaderEddineOuaich ! 😃

**Comment by LovelySanta** [comment]
That PR only solves the A-Z keys, however, I myself use usualy azerty, while I do have experience typing blind qwerty on an azerty keyboard. I do know that the key layout of the alternative characters on the keys 0-9 and the other punctuation keys are also different, so I am not sure this does solve it fully.

**Comment by T4ouf** [comment]
You are right ! I don't know really well QWERTY keyboard and I only thought about letters... I'll try to find a solution for handling punctuation keys ! (I think punctuation should be OK to implement)

However I have checked, the keys 0-9 are in the same place but creates another problem... My question is : For digit, should we try to get the number or the value of the key ? (I know that on AZERTY keyboards, your only way to access '\' is to press the "ALTGR + 8" combination and it will be recognized as entering key 8.)

=> What should do in that case ? (Create events for both ? Polling both keycodes ? Not doing anything ?)

**Comment by T4ouf** [comment]
Ok I have currently a solution for all the keys except for the digit issue (That I mentioned above). However If we consider that digit should be a priority over the other keys, it would only block 4 possible keycodes for AZERTY (defined in KeyCodes.h) : 
```cpp
		LeftBracket         = 91,  /* [ */
		Backslash           = 92,  /* \ */
		RightBracket        = 93,  /* ] */
		GraveAccent         = 96,  /* ` */
```
Currently there are no keycode for special char like '&' or '('. I think it is very AZERTY specific and I am not sure that there is a real solution to that special keys issue (at least currently with glfw). In that regard, I think we should go and accept that we can't correctly handle those 4 keycodes only for AZERTY. (It doesn't solve completely the issue but still provides a major improvment over no support at all).

Let me know what you think of it ! 😃 (If it is fine for you, I'll adapt my PR as soon as possible !)

**Comment by LovelySanta** [comment]
> Currently there are no keycode for special char like '&' or '('. I think it is very AZERTY specific and I am not sure that there is a real solution to that special keys issue (at least currently with glfw). In that regard, I think we should go and accept that we can't correctly handle those 4 keycodes only for AZERTY. (It doesn't solve completely the issue but still provides a major improvment over no support at all).

Yes there are. So on azerty, the setup is when you press the `1` key, you get an & character. When you want a 1 character, you press shift + `1`. 

The code doesn't look bad, but I feel like this might need a proper test. I think once we have some text rendering that we can make a test setup for this where you can select the different key layouts from a menu, which will then raise the correct events, and type the corresponding letters on the screen.

A quick test with the current hazel, not this PR revealed that I press the letter `A` on my azerty keyboard, and it does raise the keycode `Q`. This means that no matter the layout, the keycodes map to a physical location, and not an actual character indeed as mentioned in the initial post.

**Comment by T4ouf** [comment]
Actually no they are not : GLFW handles **each key** separately. In the case with numbers (like 1) "Shift+1" will only result in 2 distinct key callbacks and then 2 events (One for 'Shift' and one for '&' on AZERTY). That is why I was mentioning that digits will be a problem.

However text rendering will not be an issue as the glfwSetCharCallback() works already fine with every layout (you can test it with the ImGui demo window : when you type text, it always gets the correct key). 

The issue is related to key events with GLFW only. In the documentation for glfwGetKey() you can read : 
`
 *  The key functions deal with physical keys, with [key tokens](@ref keys)
 *  named after their use on the standard US keyboard layout.  If you want to
 *  input text, use the Unicode character callback instead.
`

That is why I tried to patch the keycode before the Hazel Event is created => We want Hazel to have correct events regardless of the keyboard layout. The problem is that for some layouts, some keys are not accessible (the 4 mentioned above for AZERTY). 

Is the problem more clear ? (I tried my best with my explanations ^^)

**Comment by salimp2009** [comment]
hi guys 
GLFW site has an issue regarding that . See if this helps;
https://github.com/glfw/glfw/issues/114

**Comment by salimp2009** [comment]
see if this helps regarding the same issue but I believe the solution was character callback
https://github.com/glfw/glfw/issues/114

**Comment by T4ouf** [comment]
From what I read, char Callback is really **not recommended** for anything else than text input (which is not the case here). Even in this glfw issue they use glfwGetName(). 

I think the question is : "Do we want to use those unusal keys as events ?". How often have you seen an application using "GraveAccent" or "Backslash" as a control key ? (I am not talking about text management because this is already correctly handled in Hazel.) 

From there, there are only 2 possibilities here : 
- Either you think it is important to keep those => Then there is no perfect management for every keyboard layout. (cf. my previous post) In this scenario, my PR solves the issue for all keys except those problematic keys I mentionned.
- Or you think these are not important keys => Then we need to remove those keys from our keycodes and adapt the code & my PR.

Either way, I think we should do something quickly ! Currently, Hazel is not usable on any other Keyboard layout than QWERTY and as the development progresses, it will become harder and harder to make a change (in my opinion).

Let me know what you think about it ! 😃
